src/Symfony/Component/Security/Core/Authentication/Provider/SimpleAuthenticationProvider.php

```diff
@@ -24,17 +26,22 @@ class SimpleAuthenticationProvider implements AuthenticationProviderInterface
     private $simpleAuthenticator;
     private $userProvider;
     private $providerKey;
+    private $userChecker;
 
-    public function __construct(SimpleAuthenticatorInterface $simpleAuthenticator, UserProviderInterface $userProvider, $providerKey)
+    public function __construct(SimpleAuthenticatorInterface $simpleAuthenticator, UserProviderInterface $userProvider, $providerKey, UserCheckerInterface $userChecker = null)
     {
         $this->simpleAuthenticator = $simpleAuthenticator;
         $this->userProvider = $userProvider;
         $this->providerKey = $providerKey;
+        $this->userChecker = $userChecker ?: new UserChecker();
     }
 
     public function authenticate(TokenInterface $token)
     {
+        $user = $this->userProvider->loadUserByUsername($token->getUsername());
+        $this->userChecker->checkPreAuth($user);
         $authToken = $this->simpleAuthenticator->authenticateToken($token, $this->userProvider, $this->providerKey);
```

Comment:
While I understand the need to have the user here, it would be loaded twice now

Reply:
@iltar maybe then skip the pre auth check since it can be done during authentication of the token.

Reply:
What if you do `$this->userChecker->checkPreAuth($authToken->getUser())`? I know it's not exactly correct flow wise, but it beats fetching the user twice :thinking:

Reply:
@iltar I was thinking about it but it sounds strange to me to pre authenticate authenticated token.
Changin interface to exclude the userProvider would cause BC breaks. 
You are correct about twice loading the user and simple authenticator could do the pre-check if needed.
Maybe best is to remove the `checkPreAuth`

Reply:
While it might seem strange, guard does it before/after the credentials check: https://github.com/symfony/symfony/blob/a8dc953279621398a4361c0beeff22c54cc9fce5/src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php#L124-L128

I don't see another option to be honest, leaving it out would be as not having it at all.